### PR TITLE
Start building a universal System.Linq.Expressions library

### PR DESCRIPTION
**Pull request by MichalStrehovsky**
System.Linq.Expressions currently offers multiple build time definitions to build different flavors of the library:

* `FEATURE_COMPILE` (defined everywhere except iOS/tvOS/Catalyst, and NativeAOT) - enables Reflection.Emit-based expression tree compiler.
* `FEATURE_INTERPRET` (always defined and not actually possible to build without) - enables the expression interpreter
* `FEATURE_DLG_INVOKE` (defined everywhere except NativeAOT, but we likely need to be able to run without it on iOS too because there's uninvestigated bugs around it ActiveIssue'd in #54970) - in the interpreter, use a delegate to invoke `CallInstructions` instead of `MethodInfo.Invoke`. The delegate might have to be Reflection.Emitted if it's not supportable with `Func`/`Action` (that's the uninvestigated iOS/tvOS/Catalyst bug).

For #61231, we need to be able to build a single System.Linq.Expression library that can switch between these build-time configurations _at publish time_ since we don't want to build a separate S.L.Expressions library for NativeAOT. There are advantages in having this setup for non-NativeAOT scenarios too. This pull request accomplishes that by mechanically changing the `#define`s into feature switches.

The feature switch is placed in the last proposed location for https://github.com/dotnet/runtime/issues/17973. I expect we'll want such API to be public at some point; now that Xamarin and NativeAOT use this formerly .NET Native-only thing the API request became relevant again.

This pull request is focused on the mechanical replacement of `#defines` with feature switches and it's already a lot bigger than I'm comfortable with.

There's some obvious "`!FEATURE_COMPILE` means this is .NET Native with everything what that meant" that I did not touch because this is meant to be a mechanical change. Some cleanup will be needed at some point. Right now this just mostly means we're running fewer tests than we could.

Validation:
* Verified that we're still running the same number of tests with CoreCLR as we previously were and they're all passing.
* Verified we're getting mostly the same size of the S.L.Expressions library on iOS (433 kB grew to ~~437~~ 436 kB, the diffs are expected).
* Verified things work on the NativeAOT side as well.

Cc @jkotas @marek-safar @MaximLipnin @steveisok @akoeplinger

### CODE REVIEW

File: src/libraries/System.Linq.Expressions/src/System/Dynamic/Utils/DelegateHelpers.cs
Comment (MichalStrehovsky) [inline src/libraries/System.Linq.Expressions/src/System/Dynamic/Utils/DelegateHelpers.cs:15]: @marek-safar @vitek-karas I had to hack around this like this because this:

* Should be a constant `true`,
* Unless overriden by a feature switch at publish time.

ILLink is propagating the constant true _at library build time_ and I can't figure out how else to prevent it from doing that during library build. It would be fine to constprop at app publish time.

Comment (marek-safar) [reply src/libraries/System.Linq.Expressions/src/System/Dynamic/Utils/DelegateHelpers.cs:15]: Alternatively you could disable constprop during libraries build

Comment (vitek-karas) [reply src/libraries/System.Linq.Expressions/src/System/Dynamic/Utils/DelegateHelpers.cs:15]: If this is driven by a feature switch, then it should follow the feature switch pattern - that is ask for the value from `AppContext`. The link substitutions may force it to a const value, but that should only be done on app publish time, not in library build.

Comment (MichalStrehovsky) [reply src/libraries/System.Linq.Expressions/src/System/Dynamic/Utils/DelegateHelpers.cs:15]: This should not be a user visible setting - JIT-based CoreCLR flavor will probably never implement the API flipping this switch is going to call; we might want to implement that API on Mono, but it's not implemented right now. This should stay `#define`-like - it should not be selected by a user; it should be selected by us.

I'll see how much other work constprop is doing during library build for this library and if it's okay to fully disable.

File: src/libraries/System.Linq.Expressions/src/ILLink/ILLink.Substitutions.xml
Comment (MichalStrehovsky) [reply src/libraries/System.Linq.Expressions/src/ILLink/ILLink.Substitutions.xml:4]: https://github.com/dotnet/runtime/issues/17973#issuecomment-277157066 has reasoning for why this needs to be different from RuntimeFeature.IsDynamicCodeSupported.

> Also the defaults don't look correct for existing configurations (e.g. published trimmed console app).

Could you clarify?

File: src/libraries/System.Linq.Expressions/src/ILLink/ILLink.Substitutions.xml
Comment (MichalStrehovsky) [reply src/libraries/System.Linq.Expressions/src/ILLink/ILLink.Substitutions.xml:11]: Property living in Expressions was the last direction https://github.com/dotnet/runtime/issues/17973#issuecomment-277144969 discussed. I'm just following that, given the API is unreviewed.

get_CanEmitObjectArrayDelegate and get_CanCreateArbitraryDelegates probably won't be APIs ever. CanCompileToIL might.

File: src/libraries/System.Linq.Expressions/src/System/Dynamic/Utils/DelegateHelpers.cs
Comment (MichalStrehovsky) [reply src/libraries/System.Linq.Expressions/src/System/Dynamic/Utils/DelegateHelpers.cs]: I made this into constants and disabled ipconsprop instead.

Disabling ipconstprop made the size worse by 512 bytes (less than 0.1%) for non-iDevice workloads. But having an actual constant there will result in better trimming at publish time.

The iOS version size got better by 512 bytes on the other hand - we won't need to switch this one between Ref.Emit and interpreter so it can keep the constprop and making these into constants allowed a tiny bit of code to be elimintated.

**Comment by MichalStrehovsky** [comment]
@MaximLipnin I can't make much sense from the runtime-manual run because it seems to have many unrelated failures. Could you please see if there's anything needing my attention?

**Comment by MaximLipnin** [comment]
> @MaximLipnin I can't make much sense from the runtime-manual run because it seems to have many unrelated failures. Could you please see if there's anything needing my attention?

@MichalStrehovsky JFYI - By default, only `System.Runtime` test suite executes for mobile platforms. The `runtime-manual` lanes run the whole set of libraries test suites (all the unrelated issue are going to be fixed soon). One can run it on every significant push (in case it can affect mobile targets). There is also a rolling build which will catch anything anyways.

File: src/libraries/System.Linq.Expressions/src/System/Dynamic/Utils/DelegateHelpers.cs
Comment (MichalStrehovsky) [reply src/libraries/System.Linq.Expressions/src/System/Dynamic/Utils/DelegateHelpers.cs]: Yes, I think you summed it up well. This pull request is basically just changing `#define`s into feature switches that are (at least initially) not exposed in user-friendly ways. The main reason is that NativeAOT needs to compile with a different set of defines. There is also an iDevice-specific substitution file that emulates the preexisting behavior of this being a compile time define for those platforms.

In the future, we could entertain an option to expose part of this as a public thing - "always use interpreter" could become a public feature switch that would have an appropriate AppContext value (to emulate when not trimming) and set the feature switch to remove expression compilation support when actually trimming. That's not in scope of this pull request but could be a follow on work item.

This pull request introduces 3 feature switches. One makes some sense to potentially expose publicly in the future (but it would be opt in, not default with trimming). one makes some marginal sense to expose publicly but is probably not worth it, and one is internal-only that we should ideally enable whenever ref-emit is fully disabled in the underlying runtime (once Mono supports that fallback API) - it doesn't make sense to individually expose that one.

Comment (vitek-karas) [reply src/libraries/System.Linq.Expressions/src/System/Dynamic/Utils/DelegateHelpers.cs]: > One thing I don't understand: This behavior effectively requires trimming to happen in order for the code to be runnable on platforms which require some of the other implementations. How is that supposed to work with debug builds. If I remember correctly even iOS builds don't run trimming for debug builds (they do run linker, but in "copy-only" mode, which will not perform substitutions).

Sorry - I'm probably slow here - how will this work if trimming with substitutions is NOT performed on the code?

Comment (MichalStrehovsky) [reply src/libraries/System.Linq.Expressions/src/System/Dynamic/Utils/DelegateHelpers.cs]: The substitution happens at repo build time here. There's a substitution file that applies at repo build for iDevices. Ipconstprop is also enabled there. There was no publish time configurability around this for iDevices and there won't be after this pull request.

File: src/libraries/System.Linq.Expressions/src/ILLink/ILLink.Substitutions.xml
Comment (marek-safar) [reply src/libraries/System.Linq.Expressions/src/ILLink/ILLink.Substitutions.xml:11]: I don't understand exactly what they try to do but it seems to me they react to runtime limitations, right? Should not this be then call to runtime to get the actual value with a feature switch on top of it?

File: src/libraries/System.Linq.Expressions/src/ILLink/ILLink.Substitutions.xml
Comment (marek-safar) [reply src/libraries/System.Linq.Expressions/src/ILLink/ILLink.Substitutions.xml:4]: >  has reasoning for why this needs to be different from

I don't see the reasoning in the comment and this mode is not new in .NET7. In .NET6 we use the same options to detect NativeAOT like setup where IsDynamicCodeSupported indicates that code generation is possible with IsDynamicCodeCompiled specifying how it such generated code processed. 

>  Could you clarify?

What does happen today when you publish trimmed console for desktops. I guess it has "CanCompileToIL" set to true and here we are setting it to false for everything.

File: src/libraries/System.Linq.Expressions/src/ILLink/ILLink.Substitutions.xml
Comment (MichalStrehovsky) [reply src/libraries/System.Linq.Expressions/src/ILLink/ILLink.Substitutions.xml:11]: `get_CanEmitObjectArrayDelegate` and `get_CanCreateArbitraryDelegates` could probably be replaced with `RuntimeFeature.IsDynamicCodeSupported` _eventually_, but it would be a behavior change and this pull request didn't intend to do behavior changes, just replace `#define`s with something that can be overriden.

Replacing `get_CanCreateArbitraryDelegates` with `IsDynamicCodeSupported` would make the iDevices version take the codepaths that correspond to `FEATURE_DLG_INVOKE` not being defined in the csproj and that's not how iDevices build was set up in #54970. If you would like me to make that change, I can do it, but I cannot investigate any possible fallouts from it because I don't work on or own iDevices. An observable difference will be slower perf when invoking method call instructions in the interpreter. An upside will be that some of the tests that were blocked in #54970 will work. Ideally all of this (configuring the library so that it properly works, and revisiting tests blocked in #54970) should happen in subsequent pull requests. I definitely plan to revisit the blocked tests.

`get_CanEmitObjectArrayDelegate` cannot be replaced with `RuntimeFeature.IsDynamicCodeSupported` at this moment because it requires a runtime API that is not implemented on Mono and would therefore break iDevices (it's an API to create a strongly typed Delegate that boxes all it's arguments, puts them into an array, dispatches to a `Func<object?[], object?>` delegate to do the actual work, and unboxes the result - doing all of that without runtime codegen).

File: src/libraries/System.Linq.Expressions/src/ILLink/ILLink.Substitutions.xml
Comment (MichalStrehovsky) [reply src/libraries/System.Linq.Expressions/src/ILLink/ILLink.Substitutions.xml:4]: > I don't see the reasoning in the comment and this mode is not new in .NET7. In .NET6 we use the same options to detect NativeAOT like setup where IsDynamicCodeSupported indicates that code generation is possible with IsDynamicCodeCompiled specifying how it such generated code processed.

We could make this so that the body of `CanCompileToIL` is implemented as `RuntimeFeature.IsDynamicCodeSupported`, but I'm a believer in separating behavioral changes from refactoring changes and this would be a behavioral change. It can happen in subsequent pull request as a thing that can be discussed separately from this refactoring.

> What does happen today when you publish trimmed console for desktops. I guess it has "CanCompileToIL" set to true and here we are setting it to false for everything.

The substitution is set as `feature="X" featurevalue="false" body="stub" value="false"` - if I'm reading IL Linker sources right, this substitution will be ignored unless someone sets feature X to false. Is that correct reading? I wasn't able to build intuition behind what `featurevalue=X featuredefault=Y` means and always defer to reading linker source code. If my interpretation is correct, this will do the right thing for trimmed apps (IL compiler will be kept).

**Comment by agocke** [comment]
@cston @jaredpar Could you take a look at this PR to change compile-time `#defines` into feature flags? This should allow us to create one copy of System.Linq.Expressions with both reflection and interpreter modes, where the behavior can be controlled at publish time. It shouldn't have any functional change for the default project.

**Comment by agocke** [comment]
@vitek-karas could you double check the runtime feature flag stuff here?

**Review by vitek-karas — APPROVED** [review]
The feature switch setup looks good. Assuming we discussed which feature switches we want already (I don't know enough to comment on that).

File: src/libraries/System.Linq.Expressions/tests/InterpreterTests.cs
Comment (cston) [inline src/libraries/System.Linq.Expressions/tests/InterpreterTests.cs:14]: >FEATURE_COMPILE [](http://example.com/codeflow?start=11&length=15)

Should the comment say `LambdaExpression.CanCompileToIL` instead, here and below, since there are no longer references to `FEATURE_COMPILE`?

File: src/libraries/System.Linq.Expressions/tests/Invoke/InvocationTests.cs
Comment (cston) [inline src/libraries/System.Linq.Expressions/tests/Invoke/InvocationTests.cs:247]: >FEATURE_COMPILE [](http://example.com/codeflow?start=30&length=15)

`LambdaExpression.CanCompileToIL` here and below?

File: src/libraries/System.Linq.Expressions/tests/Lambda/LambdaTests.cs
Comment (cston) [inline src/libraries/System.Linq.Expressions/tests/Lambda/LambdaTests.cs:271]: >FEATURE_COMPILE [](http://example.com/codeflow?start=53&length=15)

`LambdaExpression.CanCompileToIL`?

**Review by cston — APPROVED** [review]

File: src/libraries/System.Linq.Expressions/tests/InterpreterTests.cs
Comment (MichalStrehovsky) [reply src/libraries/System.Linq.Expressions/tests/InterpreterTests.cs]: Fixed those all.

Once this is merged, I'll do a test pass on CoreCLR with the IL Compiler disabled and see whether these comments are correct. If the comment is correct, I'll change the condition from "is IL compiler available" to "is Reflection.Emit available" and delete the comment, because that's what we're after.

**Pull request merged**
